proxy/grpcproxy/register.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcproxy
+
+import (
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/coreos/etcd/clientv3/naming"
+	gnaming "google.golang.org/grpc/naming"
+)
+
+// Register registers itself as a grpc-proxy server by writing prefixed-key
+// with a 60-second TTL-ed session. The returned channel is closed when
+// the client's context is canceled.
+func Register(c *clientv3.Client, prefix string, addr string) <-chan struct{} {
+	// write once, with default TTL 60-second
+	ss, err := concurrency.NewSession(c)
```

Comment:
this should be in a loop in the goroutine so that if the proxy is partitioned / its session expires, it can rejoin the cluster group when it reconnects